lib/Command/HSMDaemon.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * @author Jörn Friedrich Dreyer <jfd@butonic.de>
+ *
+ * @copyright Copyright (c) 2017, ownCloud GmbH
+ * @license AGPL-3.0
+ *
+ * This code is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License, version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Encryption\Command;
+
+use OCA\Encryption\AppInfo\Application;
+use OCA\Encryption\JWT;
+use OCA\Encryption\KeyManager;
+use OCP\Encryption\IEncryptionModule;
+use OCP\Http\Client\IClient;
+use OCP\Http\Client\IClientService;
+use OCP\IConfig;
+use OCP\ILogger;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class HSMDaemon extends Command {
+
+	/** @var IClient */
+	private $httpClient;
+	/** @var IConfig */
+	private $config;
+	/** @var  ILogger */
+	private $logger;
+
+	/**
+	 * @param IEncryptionModule $encryption
+	 * @param IClient $httpClient
+	 * @param IConfig $config
+	 * @param ILogger $logger
+	 */
+	public function __construct(IClientService $httpClient,
+								IConfig $config,
+								ILogger $logger) {
+		$this->httpClient = $httpClient->newClient();
+		$this->config = $config;
+		$this->logger = $logger;
+		parent::__construct();
+	}
+
+	// TODO add route for hsmdaemon to post current secret
+	// TODO add encrypt masterkey command / as option
+	// TODO add decrypt option
```

Comment:
after the merge consider raising tickets

@butonic assuming these are not critical to have in the first version ?

Reply:
correct:
- add route for hsmdaemon to post current secret
  - kill this ... I had the idea the hsmdaemon could have a command that would set up hsm based encryption and set a random shared secret. nice to have, really. Although ... it would make deployments more secure and easier to do.
- add encrypt masterkey command / as option
  - with a hsm there are actually two ways of implementing encryption:
    1. use the hsm to decrypt file individual keys
    2. use the hsm to decrypt the master key, keep it in memory (maybe even cache it with apcu), do file encryption internally
    The current implementation does the former. But we already ran into cases where every decrypt operation had to be paid for. In those cases the idea is to protect against the key being stored physically on the same machine. And the tradeof of storing it in memory vs keeping it out of the machine is driven by cost factors. Can be done in a subsequent PR. talk to @CaptionF for details.
- add decrypt option
  - this is mainly for testing and to double check the hsm setup works correctly. maybe an encrypt command makes sense as well so that the admin can try the setup on the cli ... which goes with seperating hsm.url from crypto.engine.

Reply:
@sharidas please raise so we can move forward

Reply:
then adjust https://github.com/owncloud/docs/issues/586

Reply:
Tickets created:
- [x] https://github.com/owncloud/encryption/issues/100
- [x] https://github.com/owncloud/encryption/issues/101
- [x] https://github.com/owncloud/encryption/issues/102
- [ ] Adjust https://github.com/owncloud/docs/issues/586